### ISSUE TITLE
"strategy.shouldEnter/shouldExit" does not behave exactly as "tradindRecord.getEntry/getExit"

### ISSUE THREAD

**Issue by nimo23** [mdeverdelhan/ta4j#153]
I guess, there is a bug in "StopGainRule". When running the strategy with a "TradingRecord" by "TradingRecord#getEntry/getExit" all works as expected. However, when using it by  "Strategy#shouldEnter/shouldExit" on each tick, then this rule is not followed! Please check it out. 

**I am wondering if Strategy#shouldEnter/shouldExit behave wrong in other Rules as well in compare to "TradingRecord#getEntry/getExit".** 

Why is the result of TradingRecord#getEntry/getExit not exactly the same as "Strategy#shouldEnter/shouldExit"?

**Comment by mdeverdelhan** [comment]
I don't understand. Please provide more elements. Try to debug the code to see if something is wrong.

**Comment by nimo23** [comment]
You can test it easily by yourself. Take a dataset (for example one week) and use a "StopGainRule" as a Exit-strategy (Entry-strategy does not matter).

Rule stopGain = new StopGainRule(closePrice, Decimal.valueOf("1.5"));

Now use the dataset with the strategy and print its result. One time with "Trading Record" and the other time with "shouldEnter/shouldExit":

```
TradingRecord past = series.run(s);

 for(Trade t: trades){
log.info("entry - {}", t.getEntry().getIndex());
log.info("exit - {}", t.getExit().getIndex());
}

```

Use the same dataset with the same strategy and use the tick of the timeseries:

```
if (s.shouldEnter(i)) {
log.info("entry - {}", i);
}

 } else if (s.shouldExit(i)){
log.info("exit - {}", i);
}
```


You can see easily that the Exit by "s.shouldExit" is never called (even the stop-gain rule should happen) while the same series with trading record "t.getExit().getIndex()" is called. "TradingRecord" makes it right (with having the appropriate trades), while "shouldExit" behaves wrong with the same dataset and the same exit-strategy (because shouldExit it not called even stopgain-rule should be fired).

**Comment by nimo23** [comment]
I analysed the TradingRecord and it tells me that **"Exit-Rule" is not called even the StopGainRule-Strategy should be fired.**

```
---------- TradingRecord ----------
11:58:14,247 INFO  [stdout:71] orders: [Order{type=BUY, index=14, price=1.0460000000000002629008122312371, amount=NaN}]
11:58:14,248 INFO  [stdout:71] buyOrders: [Order{type=BUY, index=14, price=1.0460000000000002629008122312371, amount=NaN}]
11:58:14,248 INFO  [stdout:71] sellOrders: []
11:58:14,248 INFO  [stdout:71] entryOrders: [Order{type=BUY, index=14, price=1.0460000000000002629008122312371, amount=NaN}]
11:58:14,248 INFO  [stdout:71] exitOrders: []
11:58:14,249 INFO  [stdout:71] trades: []
11:58:14,249 INFO  [stdout:71] startingType: BUY
11:58:14,249 INFO  [stdout:71] currentTrade: Entry: Order{type=BUY, index=14, price=1.0460000000000002629008122312371, amount=NaN} exit: null
```

**Comment by team172011** [comment]
`satisfied = currentPrice.isGreaterThanOrEqual(entryPrice.multipliedBy(gainRatioThreshold));`
To solve your problem (at the moment) you could simple change StopLossRule with StopGainRule

**Comment by mdeverdelhan** [comment]
Fixed by #172
